// gatsby-transformer-remark/extend-node-type.js
"use strict";

var _regenerator = require("babel-runtime/regenerator");

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require("babel-runtime/helpers/asyncToGenerator");

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _require = require(`graphql`),
    GraphQLObjectType = _require.GraphQLObjectType,
    GraphQLList = _require.GraphQLList,
    GraphQLString = _require.GraphQLString,
    GraphQLInt = _require.GraphQLInt,
    GraphQLEnumType = _require.GraphQLEnumType;

var Remark = require(`remark`);
var select = require(`unist-util-select`);
var sanitizeHTML = require(`sanitize-html`);
var _ = require(`lodash`);
var visit = require(`unist-util-visit`);
var toHAST = require(`mdast-util-to-hast`);
var hastToHTML = require(`hast-util-to-html`);
var mdastToToc = require(`mdast-util-toc`);
var Promise = require(`bluebird`);
var prune = require(`underscore.string/prune`);
var unified = require(`unified`);
var parse = require(`remark-parse`);
var stringify = require(`remark-stringify`);
var english = require(`retext-english`);
var remark2retext = require(`remark-retext`);
var GraphQlJson = require(`graphql-type-json`);
var stripPosition = require(`unist-util-remove-position`);
var hastReparseRaw = require(`hast-util-raw`);

var pluginsCacheStr = ``;
var pathPrefixCacheStr = ``;
var astCacheKey = function astCacheKey(node) {
  return `transformer-remark-markdown-ast-${node.internal.contentDigest}-${pluginsCacheStr}-${pathPrefixCacheStr}`;
};
var htmlCacheKey = function htmlCacheKey(node) {
  return `transformer-remark-markdown-html-${node.internal.contentDigest}-${pluginsCacheStr}-${pathPrefixCacheStr}`;
};
var htmlAstCacheKey = function htmlAstCacheKey(node) {
  return `transformer-remark-markdown-html-ast-${node.internal.contentDigest}-${pluginsCacheStr}-${pathPrefixCacheStr}`;
};
var headingsCacheKey = function headingsCacheKey(node) {
  return `transformer-remark-markdown-headings-${node.internal.contentDigest}-${pluginsCacheStr}-${pathPrefixCacheStr}`;
};
var tableOfContentsCacheKey = function tableOfContentsCacheKey(node) {
  return `transformer-remark-markdown-toc-${node.internal.contentDigest}-${pluginsCacheStr}-${pathPrefixCacheStr}`;
};

// ensure only one `/` in new url
var withPathPrefix = function withPathPrefix(url, pathPrefix) {
  return (pathPrefix + url).replace(/\/\//, `/`);
};

/**
 * Map that keeps track of generation of AST to not generate it multiple
 * times in parallel.
 *
 * @type {Map<string,Promise>}
 */
var ASTPromiseMap = new Map();

module.exports = function (_ref, pluginOptions) {
  var type = _ref.type,
      store = _ref.store,
      pathPrefix = _ref.pathPrefix,
      getNode = _ref.getNode,
      cache = _ref.cache,
      reporter = _ref.reporter;

  if (type.name !== `MarkdownRemark`) {
    return {};
  }

  pluginsCacheStr = pluginOptions.plugins.map(function (p) {
    return p.name;
  }).join(``);
  pathPrefixCacheStr = pathPrefix || ``;

  return new Promise(function (resolve, reject) {
    var getAST = function () {
      var _ref3 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(markdownNode) {
        var _this = this;

        var cacheKey, cachedAST, ASTGenerationPromise;
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                cacheKey = astCacheKey(markdownNode);
                _context2.next = 3;
                return cache.get(cacheKey);

              case 3:
                cachedAST = _context2.sent;

                if (!cachedAST) {
                  _context2.next = 8;
                  break;
                }

                return _context2.abrupt("return", cachedAST);

              case 8:
                if (!ASTPromiseMap.has(cacheKey)) {
                  _context2.next = 14;
                  break;
                }

                _context2.next = 11;
                return ASTPromiseMap.get(cacheKey);

              case 11:
                return _context2.abrupt("return", _context2.sent);

              case 14:
                ASTGenerationPromise = new Promise(function () {
                  var _ref4 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(resolve) {
                    var files, ast;
                    return _regenerator2.default.wrap(function _callee$(_context) {
                      while (1) {
                        switch (_context.prev = _context.next) {
                          case 0:
                            files = _.values(store.getState().nodes).filter(function (n) {
                              return n.internal.type === `File`;
                            });
                            _context.next = 3;
                            return new Promise(function (resolve, reject) {
                              // Use Bluebird's Promise function "each" to run remark plugins serially.
                              Promise.each(pluginOptions.plugins, function (plugin) {
                                var requiredPlugin = require(plugin.resolve);
                                if (_.isFunction(requiredPlugin.mutateSource)) {
                                  return requiredPlugin.mutateSource({
                                    markdownNode,
                                    files,
                                    getNode,
                                    reporter
                                  }, plugin.pluginOptions);
                                } else {
                                  return Promise.resolve();
                                }
                              }).then(function () {
                                var markdownAST = remark.parse(markdownNode.internal.content);

                                if (pathPrefix) {
                                  // Ensure relative links include `pathPrefix`
                                  visit(markdownAST, `link`, function (node) {
                                    if (node.url && node.url.startsWith(`/`) && !node.url.startsWith(`//`)) {
                                      node.url = withPathPrefix(node.url, pathPrefix);
                                    }
                                  });
                                }

                                // source => parse (can order parsing for dependencies) => typegen
                                //
                                // source plugins identify nodes, provide id, initial parse, know
                                // when nodes are created/removed/deleted
                                // get passed cached DataTree and return list of clean and dirty nodes.
                                // Also get passed `dirtyNodes` function which they can call with an array
                                // of node ids which will then get re-parsed and the inferred schema
                                // recreated (if inferring schema gets too expensive, can also
                                // cache the schema until a query fails at which point recreate the
                                // schema).
                                //
                                // parse plugins take data from source nodes and extend it, never mutate
                                // it. Freeze all nodes once done so typegen plugins can't change it
                                // this lets us save off the DataTree at that point as well as create
                                // indexes.
                                //
                                // typegen plugins identify further types of data that should be lazily
                                // computed due to their expense, or are hard to infer graphql type
                                // (markdown ast), or are need user input in order to derive e.g.
                                // markdown headers or date fields.
                                //
                                // wrap all resolve functions to (a) auto-memoize and (b) cache to disk any
                                // resolve function that takes longer than ~10ms (do research on this
                                // e.g. how long reading/writing to cache takes), and (c) track which
                                // queries are based on which source nodes. Also if connection of what
                                // which are always rerun if their underlying nodes change..
                                //
                                // every node type in DataTree gets a schema type automatically.
                                // typegen plugins just modify the auto-generated types to add derived fields
                                // as well as computationally expensive fields.
                                var files = _.values(store.getState().nodes).filter(function (n) {
                                  return n.internal.type === `File`;
                                });
                                // Use Bluebird's Promise function "each" to run remark plugins serially.
                                Promise.each(pluginOptions.plugins, function (plugin) {
                                  var requiredPlugin = require(plugin.resolve);
                                  if (_.isFunction(requiredPlugin)) {
                                    return requiredPlugin({
                                      markdownAST,
                                      markdownNode,
                                      getNode,
                                      files,
                                      pathPrefix,
                                      reporter
                                    }, plugin.pluginOptions);
                                  } else {
                                    return Promise.resolve();
                                  }
                                }).then(function () {
                                  resolve(markdownAST);
                                });
                              });
                            });

                          case 3:
                            ast = _context.sent;


                            // Save new AST to cache and return
                            cache.set(cacheKey, ast);
                            // We can now release promise, as we cached result
                            ASTPromiseMap.delete(cacheKey);
                            return _context.abrupt("return", resolve(ast));

                          case 7:
                          case "end":
                            return _context.stop();
                        }
                      }
                    }, _callee, _this);
                  }));

                  return function (_x2) {
                    return _ref4.apply(this, arguments);
                  };
                }());

                ASTPromiseMap.set(cacheKey, ASTGenerationPromise);
                _context2.next = 18;
                return ASTGenerationPromise;

              case 18:
                return _context2.abrupt("return", _context2.sent);

              case 19:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      return function getAST(_x) {
        return _ref3.apply(this, arguments);
      };
    }();

    var getHeadings = function () {
      var _ref5 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(markdownNode) {
        var cachedHeadings, ast, headings;
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return cache.get(headingsCacheKey(markdownNode));

              case 2:
                cachedHeadings = _context3.sent;

                if (!cachedHeadings) {
                  _context3.next = 7;
                  break;
                }

                return _context3.abrupt("return", cachedHeadings);

              case 7:
                _context3.next = 9;
                return getAST(markdownNode);

              case 9:
                ast = _context3.sent;
                headings = select(ast, `heading`).map(function (heading) {
                  return {
                    value: _.first(select(heading, `text`).map(function (text) {
                      return text.value;
                    })),
                    depth: heading.depth
                  };
                });


                cache.set(headingsCacheKey(markdownNode), headings);
                return _context3.abrupt("return", headings);

              case 13:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      return function getHeadings(_x3) {
        return _ref5.apply(this, arguments);
      };
    }();

    var getTableOfContents = function () {
      var _ref6 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee4(markdownNode) {
        var cachedToc, ast, tocAst, toc, addSlugToUrl;
        return _regenerator2.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return cache.get(tableOfContentsCacheKey(markdownNode));

              case 2:
                cachedToc = _context4.sent;

                if (!cachedToc) {
                  _context4.next = 7;
                  break;
                }

                return _context4.abrupt("return", cachedToc);

              case 7:
                _context4.next = 9;
                return getAST(markdownNode);

              case 9:
                ast = _context4.sent;
                tocAst = mdastToToc(ast);
                toc = void 0;

                if (tocAst.map) {
                  addSlugToUrl = function addSlugToUrl(node) {
                    if (node.url) {
                      node.url = [pathPrefix, markdownNode.fields.slug, node.url].join(`/`).replace(/\/\//g, `/`);
                    }
                    if (node.children) {
                      node.children = node.children.map(function (node) {
                        return addSlugToUrl(node);
                      });
                    }

                    return node;
                  };

                  tocAst.map = addSlugToUrl(tocAst.map);

                  toc = hastToHTML(toHAST(tocAst.map));
                } else {
                  toc = ``;
                }
                cache.set(tableOfContentsCacheKey(markdownNode), toc);
                return _context4.abrupt("return", toc);

              case 15:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      return function getTableOfContents(_x4) {
        return _ref6.apply(this, arguments);
      };
    }();

    var getHTMLAst = function () {
      var _ref7 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee5(markdownNode) {
        var cachedAst, ast, htmlAst;
        return _regenerator2.default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return cache.get(htmlAstCacheKey(markdownNode));

              case 2:
                cachedAst = _context5.sent;

                if (!cachedAst) {
                  _context5.next = 7;
                  break;
                }

                return _context5.abrupt("return", cachedAst);

              case 7:
                _context5.next = 9;
                return getAST(markdownNode);

              case 9:
                ast = _context5.sent;
                htmlAst = toHAST(ast, { allowDangerousHTML: true });

                // Save new HTML AST to cache and return

                cache.set(htmlAstCacheKey(markdownNode), htmlAst);
                return _context5.abrupt("return", htmlAst);

              case 13:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      return function getHTMLAst(_x5) {
        return _ref7.apply(this, arguments);
      };
    }();

    var getHTML = function () {
      var _ref8 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee6(markdownNode) {
        var cachedHTML, ast, html;
        return _regenerator2.default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return cache.get(htmlCacheKey(markdownNode));

              case 2:
                cachedHTML = _context6.sent;

                if (!cachedHTML) {
                  _context6.next = 7;
                  break;
                }

                return _context6.abrupt("return", cachedHTML);

              case 7:
                _context6.next = 9;
                return getHTMLAst(markdownNode);

              case 9:
                ast = _context6.sent;

                // Save new HTML to cache and return
                html = hastToHTML(ast, {
                  allowDangerousHTML: true
                });

                // Save new HTML to cache and return

                cache.set(htmlCacheKey(markdownNode), html);
                return _context6.abrupt("return", html);

              case 13:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      return function getHTML(_x6) {
        return _ref8.apply(this, arguments);
      };
    }();

    // Setup Remark.
    var remark = new Remark().data(`settings`, {
      commonmark: true,
      footnotes: true,
      pedantic: true
    });

    for (var _iterator = pluginOptions.plugins, _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : _iterator[Symbol.iterator]();;) {
      var _ref2;

      if (_isArray) {
        if (_i >= _iterator.length) break;
        _ref2 = _iterator[_i++];
      } else {
        _i = _iterator.next();
        if (_i.done) break;
        _ref2 = _i.value;
      }

      var plugin = _ref2;

      var requiredPlugin = require(plugin.resolve);
      if (_.isFunction(requiredPlugin.setParserPlugins)) {
        for (var _iterator2 = requiredPlugin.setParserPlugins(plugin.pluginOptions), _isArray2 = Array.isArray(_iterator2), _i2 = 0, _iterator2 = _isArray2 ? _iterator2 : _iterator2[Symbol.iterator]();;) {
          var _ref11;

          if (_isArray2) {
            if (_i2 >= _iterator2.length) break;
            _ref11 = _iterator2[_i2++];
          } else {
            _i2 = _iterator2.next();
            if (_i2.done) break;
            _ref11 = _i2.value;
          }

          var parserPlugin = _ref11;

          if (_.isArray(parserPlugin)) {
            var parser = parserPlugin[0],
                options = parserPlugin[1];

            remark = remark.use(parser, options);
          } else {
            remark = remark.use(parserPlugin);
          }
        }
      }
    }

    var HeadingType = new GraphQLObjectType({
      name: `MarkdownHeading`,
      fields: {
        value: {
          type: GraphQLString,
          resolve(heading) {
            return heading.value;
          }
        },
        depth: {
          type: GraphQLInt,
          resolve(heading) {
            return heading.depth;
          }
        }
      }
    });

    var HeadingLevels = new GraphQLEnumType({
      name: `HeadingLevels`,
      values: {
        h1: { value: 1 },
        h2: { value: 2 },
        h3: { value: 3 },
        h4: { value: 4 },
        h5: { value: 5 },
        h6: { value: 6 }
      }
    });

    return resolve({
      html: {
        type: GraphQLString,
        resolve(markdownNode) {
          return getHTML(markdownNode);
        }
      },
      htmlAst: {
        type: GraphQlJson,
        resolve(markdownNode) {
          return getHTMLAst(markdownNode).then(function (ast) {
            var strippedAst = stripPosition(_.clone(ast), true);
            return hastReparseRaw(strippedAst);
          });
        }
      },
      excerpt: {
        type: GraphQLString,
        args: {
          pruneLength: {
            type: GraphQLInt,
            defaultValue: 140
          }
        },
        resolve(markdownNode, _ref9) {
          var pruneLength = _ref9.pruneLength;

          if (markdownNode.excerpt) {
            return Promise.resolve(markdownNode.excerpt);
          }
          return getAST(markdownNode).then(function (ast) {
            var excerptNodes = [];
            visit(ast, function (node) {
              if (node.type === `text` || node.type === `inlineCode`) {
                excerptNodes.push(node.value);
              }
              return;
            });

            return prune(excerptNodes.join(` `), pruneLength, `…`);
          });
        }
      },
      headings: {
        type: new GraphQLList(HeadingType),
        args: {
          depth: {
            type: HeadingLevels
          }
        },
        resolve(markdownNode, _ref10) {
          var depth = _ref10.depth;

          return getHeadings(markdownNode).then(function (headings) {
            if (typeof depth === `number`) {
              headings = headings.filter(function (heading) {
                return heading.depth === depth;
              });
            }
            return headings;
          });
        }
      },
      timeToRead: {
        type: GraphQLInt,
        resolve(markdownNode) {
          return getHTML(markdownNode).then(function (html) {
            var timeToRead = 0;
            var pureText = sanitizeHTML(html, { allowTags: [] });
            var avgWPM = 265;
            var wordCount = _.words(pureText).length;
            timeToRead = Math.round(wordCount / avgWPM);
            if (timeToRead === 0) {
              timeToRead = 1;
            }
            return timeToRead;
          });
        }
      },
      tableOfContents: {
        type: GraphQLString,
        resolve(markdownNode) {
          return getTableOfContents(markdownNode);
        }
      },
      // TODO add support for non-latin languages https://github.com/wooorm/remark/issues/251#issuecomment-296731071
      wordCount: {
        type: new GraphQLObjectType({
          name: `wordCount`,
          fields: {
            paragraphs: {
              type: GraphQLInt
            },
            sentences: {
              type: GraphQLInt
            },
            words: {
              type: GraphQLInt
            }
          }
        }),
        resolve(markdownNode) {
          var counts = {};

          unified().use(parse).use(remark2retext, unified().use(english).use(count)).use(stringify).processSync(markdownNode.internal.content);

          return {
            paragraphs: counts.ParagraphNode,
            sentences: counts.SentenceNode,
            words: counts.WordNode
          };

          function count() {
            return counter;
            function counter(tree) {
              visit(tree, visitor);
              function visitor(node) {
                counts[node.type] = (counts[node.type] || 0) + 1;
              }
            }
          }
        }
      }
    });
  });
};
