// common-tags/lib/tags/tags.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

exports.default = tags;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function tags(opts) {
  var settings = (0, _extends3.default)({
    trim: true,
    oneLine: false,
    stripIndent: false,
    includeArrays: false
  }, opts);
  // return a tag function that transforms our template
  return function tag(template) {
    for (var _len = arguments.length, expressions = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      expressions[_key - 1] = arguments[_key];
    }

    // join the parts necessary to re-construct the template
    var temp = template.reduce(function (accumulator, part, i) {
      var expression = expressions[i - 1];
      if (settings.includeArrays && Array.isArray(expression)) {
        var sep = settings.includeArrays.separator || '';
        var con = settings.includeArrays.conjunction;
        // inline arrays, making sure to include item separator
        expression = expression.join(sep + accumulator.match(/(\s+)$/)[1]);
        if (con) {
          // replace the last separator with the conjunction
          var sepIndex = expression.lastIndexOf(sep);
          expression = expression.substr(0, sepIndex) + ' ' + con + expression.substr(sepIndex + 1);
        }
      }
      return accumulator + expression + part;
    });
    // replace any newlines with spaces if we just want
    // a one liner
    if (settings.oneLine) {
      temp = temp.replace(/(?:\s+)/g, ' ');
    }
    if (settings.oneLineTrim) {
      temp = temp.replace(/(?:\n\s+)/g, '');
    }
    if (settings.stripIndent) {
      // strip leading indents
      var match = temp.match(/^[ \t]*(?=\S)/gm);
      var indent = Math.min.apply(Math, (0, _toConsumableArray3.default)(match.map(function (el) {
        return el.length;
      })));
      var regexp = new RegExp('^[ \\t]{' + indent + '}', 'gm');
      temp = indent > 0 ? temp.replace(regexp, '') : temp;
    }
    // trim leading and trailing whitespace
    if (settings.trim) {
      temp = temp.trim();
    }
    return temp;
  };
}
