// gatsby/dist/schema/build-connection-fields.js
"use strict";

var _extends2 = require("babel-runtime/helpers/extends");

var _extends3 = _interopRequireDefault(_extends2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _ = require(`lodash`);

var _require = require(`graphql`),
    GraphQLInt = _require.GraphQLInt,
    GraphQLList = _require.GraphQLList,
    GraphQLString = _require.GraphQLString,
    GraphQLEnumType = _require.GraphQLEnumType;

var _require2 = require(`graphql-skip-limit`),
    connectionArgs = _require2.connectionArgs,
    connectionDefinitions = _require2.connectionDefinitions,
    connectionFromArray = _require2.connectionFromArray;

var _require3 = require(`./data-tree-utils`),
    buildFieldEnumValues = _require3.buildFieldEnumValues;

module.exports = function (type) {
  var enumValues = buildFieldEnumValues({
    nodes: type.nodes,
    typeName: type.name
  });

  var _connectionDefinition = connectionDefinitions({
    name: _.camelCase(`${type.name} groupConnection`),
    nodeType: type.nodeObjectType,
    connectionFields: function connectionFields() {
      return {
        field: { type: GraphQLString },
        fieldValue: { type: GraphQLString },
        totalCount: { type: GraphQLInt }
      };
    }
  }),
      groupConnection = _connectionDefinition.connectionType;

  return {
    totalCount: {
      type: GraphQLInt
    },
    distinct: {
      type: new GraphQLList(GraphQLString),
      args: {
        field: {
          type: new GraphQLEnumType({
            name: _.camelCase(`${type.name} distinct enum`),
            values: enumValues
          })
        }
      },
      resolve(connection, args) {
        var fieldName = args.field;
        if (_.includes(args.field, `___`)) {
          fieldName = args.field.replace(`___`, `.`);
        }
        var fields = connection.edges.map(function (edge) {
          return _.get(edge.node, fieldName);
        });
        return _.sortBy(_.filter(_.uniq(_.flatten(fields)), _.identity));
      }
    },
    group: {
      type: new GraphQLList(groupConnection),
      args: (0, _extends3.default)({}, connectionArgs, {
        field: {
          type: new GraphQLEnumType({
            name: _.camelCase(`${type.name} group enum`),
            values: enumValues
          })
        }
      }),
      resolve(connection, args) {
        var fieldName = args.field.replace(`___`, `.`);
        var connectionNodes = connection.edges.map(function (edge) {
          return edge.node;
        });

        var groups = {};
        // Do a custom grouping for arrays (w/ a group per array value)
        // Find the first node with this field and check if it's an array.
        if (_.isArray(_.get(_.find(connectionNodes, fieldName), fieldName))) {
          var values = _.uniq(_.reduce(connectionNodes, function (vals, n) {
            if (_.has(n, fieldName)) {
              return vals.concat(_.get(n, fieldName));
            } else {
              return vals;
            }
          }, []));
          values.forEach(function (val) {
            groups[val] = _.filter(connectionNodes, function (n) {
              return _.includes(_.get(n, fieldName), val);
            });
          });
        } else {
          groups = _.groupBy(connectionNodes, fieldName);
        }
        var groupConnections = [];

        // Do default sort by fieldValue
        var sortedFieldValues = _.sortBy(_.keys(groups));
        _.each(sortedFieldValues, function (fieldValue) {
          var groupNodes = groups[fieldValue];
          var groupConn = connectionFromArray(groupNodes, args);
          groupConn.totalCount = groupNodes.length;
          groupConn.field = fieldName;
          groupConn.fieldValue = fieldValue;
          groupConnections.push(groupConn);
        });

        return groupConnections;
      }
    }
  };
};
