// gatsby/dist/schema/infer-graphql-input-fields.js
"use strict";

exports.__esModule = true;

var _extends2 = require("babel-runtime/helpers/extends");

var _extends3 = _interopRequireDefault(_extends2);

exports.inferInputObjectStructureFromNodes = inferInputObjectStructureFromNodes;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _require = require(`graphql`),
    GraphQLInputObjectType = _require.GraphQLInputObjectType,
    GraphQLBoolean = _require.GraphQLBoolean,
    GraphQLString = _require.GraphQLString,
    GraphQLFloat = _require.GraphQLFloat,
    GraphQLInt = _require.GraphQLInt,
    GraphQLList = _require.GraphQLList;

var _require2 = require(`common-tags`),
    oneLine = _require2.oneLine;

var _ = require(`lodash`);
var invariant = require(`invariant`);
var typeOf = require(`type-of`);
var createTypeName = require(`./create-type-name`);
var createKey = require(`./create-key`);

var _require3 = require(`./data-tree-utils`),
    getExampleValues = _require3.getExampleValues,
    extractFieldNames = _require3.extractFieldNames,
    isEmptyObjectOrArray = _require3.isEmptyObjectOrArray;

var _require4 = require(`./infer-graphql-type`),
    findLinkedNode = _require4.findLinkedNode;

var _require5 = require(`../redux`),
    getNodes = _require5.getNodes;

function typeFields(type) {
  switch (type) {
    case `boolean`:
      return {
        eq: { type: GraphQLBoolean },
        ne: { type: GraphQLBoolean }
      };
    case `string`:
      return {
        eq: { type: GraphQLString },
        ne: { type: GraphQLString },
        regex: { type: GraphQLString },
        glob: { type: GraphQLString }
      };
    case `int`:
      return {
        eq: { type: GraphQLInt },
        ne: { type: GraphQLInt },
        gt: { type: GraphQLInt },
        gte: { type: GraphQLInt },
        lt: { type: GraphQLInt },
        lte: { type: GraphQLInt }
      };
    case `float`:
      return {
        eq: { type: GraphQLFloat },
        ne: { type: GraphQLFloat },
        gt: { type: GraphQLFloat },
        gte: { type: GraphQLFloat },
        lt: { type: GraphQLFloat },
        lte: { type: GraphQLFloat }
      };
  }
  return {};
}

function inferGraphQLInputFields(_ref) {
  var value = _ref.value,
      nodes = _ref.nodes,
      prefix = _ref.prefix;

  if (value == null || isEmptyObjectOrArray(value)) return null;

  switch (typeOf(value)) {
    case `array`:
      {
        var headValue = value[0];
        var headType = typeOf(headValue);

        if (headType === `number`) headType = _.isInteger(headValue) ? `int` : `float`;

        // Determine type for in operator.
        var inType = void 0;
        switch (headType) {
          case `int`:
            inType = GraphQLInt;
            break;
          case `float`:
            inType = GraphQLFloat;
            break;
          case `date`:
          case `string`:
            inType = GraphQLString;
            break;
          case `boolean`:
            inType = GraphQLBoolean;
            break;
          case `array`:
          case `object`:
            {
              var inferredField = inferGraphQLInputFields({
                value: headValue,
                prefix,
                nodes
              });
              invariant(inferredField, `Could not infer graphQL type for value: ${JSON.stringify(Object.keys(headValue))}`);
              inType = inferredField.type;
              break;
            }
          default:
            invariant(false, oneLine`
              Could not infer an appropriate GraphQL input type
              for value: ${headValue} of type ${headType} along path: ${prefix}
            `);
        }

        return {
          type: new GraphQLInputObjectType({
            name: createTypeName(`${prefix}QueryList`),
            fields: (0, _extends3.default)({}, typeFields(headType), {
              in: { type: new GraphQLList(inType) }
            })
          })
        };
      }
    case `boolean`:
      {
        return {
          type: new GraphQLInputObjectType({
            name: createTypeName(`${prefix}QueryBoolean`),
            fields: typeFields(`boolean`)
          })
        };
      }
    case `date`:
    case `string`:
      {
        return {
          type: new GraphQLInputObjectType({
            name: createTypeName(`${prefix}QueryString`),
            fields: typeFields(`string`)
          })
        };
      }
    case `object`:
      {
        var fields = inferInputObjectStructureFromNodes({
          nodes,
          prefix,
          exampleValue: value
        }).inferredFields;
        if (!_.isEmpty(fields)) {
          return {
            type: new GraphQLInputObjectType({
              name: createTypeName(`${prefix}InputObject`),
              fields
            })
          };
        } else {
          return null;
        }
      }
    case `number`:
      {
        if (value % 1 === 0) {
          return {
            type: new GraphQLInputObjectType({
              name: createTypeName(`${prefix}QueryInteger`),
              fields: typeFields(`int`)
            })
          };
        } else {
          return {
            type: new GraphQLInputObjectType({
              name: createTypeName(`${prefix}QueryFloat`),
              fields: typeFields(`float`)
            })
          };
        }
      }
    default:
      return null;
  }
}

var EXCLUDE_KEYS = {
  parent: 1,
  children: 1
};

var recursiveOmitBy = function recursiveOmitBy(value, fn) {
  if (_.isObject(value)) {
    if (_.isPlainObject(value)) {
      value = _.omitBy(value, fn);
    }
    _.each(value, function (v, k) {
      value[k] = recursiveOmitBy(v, fn);
    });
    if (_.isEmpty(value)) {
      // don't return empty objects - gatsby doesn't support these
      return null;
    }
  }
  return value;
};

var linkedNodeCache = {};

function inferInputObjectStructureFromNodes(_ref2) {
  var nodes = _ref2.nodes,
      _ref2$typeName = _ref2.typeName,
      typeName = _ref2$typeName === undefined ? `` : _ref2$typeName,
      _ref2$prefix = _ref2.prefix,
      prefix = _ref2$prefix === undefined ? `` : _ref2$prefix,
      _ref2$exampleValue = _ref2.exampleValue,
      exampleValue = _ref2$exampleValue === undefined ? null : _ref2$exampleValue;

  var inferredFields = {};
  var isRoot = !prefix;

  prefix = isRoot ? typeName : prefix;
  if (exampleValue === null) {
    exampleValue = getExampleValues({ nodes, typeName });
  }

  _.each(exampleValue, function (v, k) {
    var value = v;
    var key = k;
    // Remove fields for traversing through nodes as we want to control
    // setting traversing up not try to automatically infer them.
    if (isRoot && EXCLUDE_KEYS[key]) return;

    if (_.includes(key, `___NODE`)) {
      // TODO: Union the objects in array
      var nodeToFind = _.isArray(value) ? value[0] : value;
      var linkedNode = findLinkedNode(nodeToFind);

      // Get from cache if found, else store into it
      if (linkedNodeCache[linkedNode.internal.type]) {
        value = linkedNodeCache[linkedNode.internal.type];
      } else {
        var relatedNodes = getNodes().filter(function (node) {
          return node.internal.type === linkedNode.internal.type;
        });
        value = getExampleValues({
          nodes: relatedNodes,
          typeName: linkedNode.internal.type
        });
        value = recursiveOmitBy(value, function (_v, _k) {
          return _.includes(_k, `___NODE`);
        });
        linkedNodeCache[linkedNode.internal.type] = value;
      }

      if (_.isArray(value)) {
        value = [value];
      }

      ;
      var _key$split = key.split(`___`);

      key = _key$split[0];
    }

    var field = inferGraphQLInputFields({
      nodes,
      value,
      prefix: `${prefix}${_.upperFirst(key)}`
    });

    if (field == null) return;
    inferredFields[createKey(key)] = field;
  });

  // Add sorting (but only to the top level).
  var sort = [];
  if (typeName) {
    sort = extractFieldNames(nodes);
  }

  return { inferredFields, sort };
}
